File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.swift
// Abstract Factory pattern in Swift.
//
// Protocols model both the products and the factory. The client is written
// as a generic function over the factory protocol, which avoids existential
// (`any`) overhead while staying fully abstract.

import Foundation

protocol Button {
    func render() -> String
}

protocol Checkbox {
    func render() -> String
}

struct MacButton: Button {
    func render() -> String { "( OK )  // rounded macOS button" }
}

struct MacCheckbox: Checkbox {
    func render() -> String { "◉ Remember me  // macOS checkbox" }
}

struct WindowsButton: Button {
    func render() -> String { "[ OK ]  // rectangular Windows button" }
}

struct WindowsCheckbox: Checkbox {
    func render() -> String { "[x] Remember me  // Windows checkbox" }
}

/// One creation method per product kind; products are guaranteed to match.
protocol WidgetFactory {
    func createButton() -> Button
    func createCheckbox() -> Checkbox
}

struct MacFactory: WidgetFactory {
    func createButton() -> Button { MacButton() }
    func createCheckbox() -> Checkbox { MacCheckbox() }
}

struct WindowsFactory: WidgetFactory {
    func createButton() -> Button { WindowsButton() }
    func createCheckbox() -> Checkbox { WindowsCheckbox() }
}

/// Client code: depends only on the factory protocol, never on a family.
func renderDialog<F: WidgetFactory>(using factory: F) {
    print(factory.createButton().render())
    print(factory.createCheckbox().render())
}

print("-- macOS theme --")
renderDialog(using: MacFactory())

print("-- Windows theme --")
renderDialog(using: WindowsFactory())

File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.rs
// Abstract Factory pattern in Rust.
//
// Traits model the products and the factory; boxed trait objects allow the
// family to be chosen at runtime. When the set of families is closed and
// known up front, an enum with match arms is often more idiomatic.

trait Button {
    fn render(&self) -> String;
}

trait Checkbox {
    fn render(&self) -> String;
}

struct MacButton;
struct MacCheckbox;
struct WindowsButton;
struct WindowsCheckbox;

impl Button for MacButton {
    fn render(&self) -> String {
        "( OK )  // rounded macOS button".to_string()
    }
}

impl Checkbox for MacCheckbox {
    fn render(&self) -> String {
        "◉ Remember me  // macOS checkbox".to_string()
    }
}

impl Button for WindowsButton {
    fn render(&self) -> String {
        "[ OK ]  // rectangular Windows button".to_string()
    }
}

impl Checkbox for WindowsCheckbox {
    fn render(&self) -> String {
        "[x] Remember me  // Windows checkbox".to_string()
    }
}

/// One creation method per product kind; products are guaranteed to match.
trait WidgetFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
}

struct MacFactory;
struct WindowsFactory;

impl WidgetFactory for MacFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(MacButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(MacCheckbox)
    }
}

impl WidgetFactory for WindowsFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(WindowsButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(WindowsCheckbox)
    }
}

/// Client code: depends only on the factory trait, never on a family.
fn render_dialog(factory: &dyn WidgetFactory) {
    println!("{}", factory.create_button().render());
    println!("{}", factory.create_checkbox().render());
}

fn main() {
    let themes: [(&str, Box<dyn WidgetFactory>); 2] = [
        ("macOS", Box::new(MacFactory)),
        ("Windows", Box::new(WindowsFactory)),
    ];

    for (name, factory) in &themes {
        println!("-- {name} theme --");
        render_dialog(factory.as_ref());
    }
}
